indexer/validate_and_format_contacts.cpp

```diff
@@ -0,0 +1,400 @@
+      return linePageClean;
+  }
+
+  if (EditableMapObject::ValidateWebsite(linePage))
+  {
+    // URL schema documentation: https://developers.line.biz/en/docs/messaging-api/using-line-url-scheme/
+
+    string linePageUrl = linePage;
+    // Check if HTTP protocol is present.
+    if (!strings::StartsWith(linePage, "http://") && !strings::StartsWith(linePage, "https://"))
+      linePageUrl = "https://" + linePage;
+
+    url::Url const url = url::Url(linePageUrl);
+    string const & domain = strings::MakeLowerCase(url.GetWebDomain());
+    // Check Line domain name.
+    if (domain == "page.line.me")
+    {
+      // Parse https://page.line.me/{LINE ID}
+      string lineId = url.GetWebPath();
+      return stripAtSymbol(lineId);
+    }
+    else if (domain == "line.me" || strings::EndsWith(domain, ".line.me"))
+    {
+      auto webPath = url.GetWebPath();
+      if (strings::StartsWith(webPath, "R/ti/p/"))
+      {
+        // Parse https://line.me/R/ti/p/{LINE ID}
+        string lineId = webPath.substr(7, webPath.length());
+        return stripAtSymbol(lineId);
+      }
+      else if (strings::StartsWith(webPath, "ti/p/"))
+      {
+        // Parse https://line.me/ti/p/{LINE ID}
+        string lineId = webPath.substr(5, webPath.length());
+        return stripAtSymbol(lineId);
+      }
+      else if (strings::StartsWith(webPath, "R/home/public/main") || strings::StartsWith(webPath, "R/home/public/profile"))
+      {
+        // Parse https://line.me/R/home/public/main?id={LINE ID without @}
+        // and https://line.me/R/home/public/profile?id={LINE ID without @}
+        string lineId = {};
+        url.ForEachParam([&lineId](url::Param const & param) {
+          if (param.m_name == "id")
+            lineId = param.m_value;
+        });
+
+        return lineId;
+      }
+      else
+      {
+        if (strings::StartsWith(linePage, "http://"))
+          return linePage.substr(7);
+        if (strings::StartsWith(linePage, "https://"))
+          return linePage.substr(8);
+      }
+    }
+  }
+
+  return {};
+}
+
+bool ValidateFacebookPage(string const & page)
+{
+  if (page.empty())
+    return true;
+
+  // See rules: https://www.facebook.com/help/105399436216001
+  if (regex_match(page, s_fbRegex))
+    return true;
+
+  if (EditableMapObject::ValidateWebsite(page))
+  {
+    string const domain = strings::MakeLowerCase(url::Url::FromString(page).GetWebDomain());
+    return (strings::StartsWith(domain, "facebook.") || strings::StartsWith(domain, "fb.") ||
+            domain.find(".facebook.") != string::npos || domain.find(".fb.") != string::npos);
+  }
+
+  return false;
+}
+
+bool ValidateInstagramPage(string const & page)
+{
+  if (page.empty())
+    return true;
+
+  // Rules took here: https://blog.jstassen.com/2016/03/code-regex-for-instagram-username-and-hashtags/
+  if (regex_match(page, s_instaRegex))
+    return true;
+
+  if (EditableMapObject::ValidateWebsite(page))
```

Comment:
nit: Negate the if, return false, and make the code shorter without bracers.

Reply:
Done